node_widget.py

```diff
@@ -0,0 +1,159 @@
+
+        #window node belonds in
+        self.relative_window = relative_window
+
+
+        #icon for the drag button
+        self.add_widget(Image(source='drag.png'))
+
+    #prepare to drag
+    def on_touch_down(self, touch):
+        if self.collide_point(*touch.pos):
+            self.node_widget.is_dragging = True
+            self.node_widget.touch_x = touch.x - self.node_widget.x
+            self.node_widget.touch_y = touch.y - self.node_widget.y
+            return True
+        return super().on_touch_down(touch)
+
+    #update movement on dragging
+    def on_touch_move(self, touch):
+        if self.node_widget.is_dragging:
+            #updates location of the mouse
+            new_x = touch.x - self.node_widget.touch_x
+            new_y = touch.y - self.node_widget.touch_y
+
+            # Check if the new position is within the boundaries of the relative window
+            if 0 <= new_x <= self.relative_window.width - self.node_widget.width:
+
+                #changes position of the connected lines
+                for line in self.node_widget.incoming_lines:
+                    in_line = line[0]
+                    points = in_line.points
+                    in_line.points = [new_x + 50, points[1], points[2], points[3]]
+                for line in self.node_widget.outgoing_lines:
+                    out_line = line[0]
+                    points = out_line.points
+                    out_line.points = [points[0], points[1], new_x + 50, points[3]]
+                
+                #changes node position
+                self.node_widget.x = new_x
+
+            if 0 <= new_y <= self.relative_window.height - self.node_widget.height:
+
+                #changes the position of connecting lines
+                for line in self.node_widget.incoming_lines:
+                    in_line = line[0]
+                    points = in_line.points
+                    in_line.points = [points[0], new_y + 50, points[2], points[3]]
+                for line in self.node_widget.outgoing_lines:
+                    out_line = line[0]
+                    points = out_line.points
+                    out_line.points = [points[0], points[1], points[2], new_y + 50]
+                
+                #changes node position
+                self.node_widget.y = new_y
+
+            return True
+        return super().on_touch_move(touch)
+
+    #finish dragging process
+    def on_touch_up(self, touch):
+        if self.node_widget.is_dragging:
+            self.node_widget.is_dragging = False
+            return True
+        return super().on_touch_up(touch)
+
+#a drag-and-drop widget used for nodes in the tree
+class DraggableWidget(BoxLayout):
+    is_dragging = False
+    touch_x = 0
+    touch_y = 0
+
+    def __init__(self, **kwargs):
+
+        #the array holders for the visual lines connecting nodes that link outputs and inputs
+        self.incoming_lines = []
+        self.outgoing_lines = []
```

Comment:
```suggestion
class DraggableWidget(BoxLayout):
    def __init__(self, **kwargs):
        self.is_dragging = False
        self.touch_x = 0
        self.touch_y = 0

        #the array holders for the visual lines connecting nodes that link outputs and inputs
        self.incoming_lines = []
        self.outgoing_lines = []
```
The more I think about it, the more I'm not a fan of using class attributes as defaults. There's a lot of possible gotchas that can come out of it without the correct security in place, and it's just simpler to move the defaults into the `__init__`.